sdk/identity/Azure.Identity/src/AzureCliCredentialClient.cs

```diff
@@ -0,0 +1,172 @@
+        private readonly string _workingDir;
+        private readonly string _path;
+
+        // The default install paths are used to find Azure CLI if no path is specified. This is to prevent executing out of the current working directory.
+        private static readonly string DefaultPathWindows = $"{EnvironmentVariables.ProgramFilesX86}\\Microsoft SDKs\\Azure\\CLI2\\wbin;{EnvironmentVariables.ProgramFiles}\\Microsoft SDKs\\Azure\\CLI2\\wbin";
+        private const string DefaultPath = "/usr/bin:/usr/local/bin";
+
+        public AzureCliCredentialClient()
+        {
+            (_path, _workingDir) = GetSafePathAndWorkingDirectory();
+        }
+
+        public virtual AccessToken RequestCliAccessToken(string[] scopes, CancellationToken cancellationToken)
+        {
+            return RequestCliAccessTokenAsync(false, scopes, cancellationToken).EnsureCompleted();
+        }
+
+        public virtual async ValueTask<AccessToken> RequestCliAccessTokenAsync(string[] scopes, CancellationToken cancellationToken)
+        {
+            return await RequestCliAccessTokenAsync(true, scopes, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected virtual async ValueTask<AccessToken> RequestCliAccessTokenAsync(bool async, string[] scopes, CancellationToken cancellationToken)
+        {
+            string resource = ScopeUtilities.ScopesToResource(scopes);
+
+            ScopeUtilities.ValidateScope(resource);
+
+            (string output, int exitCode) = await GetAzureCliAccesToken(async, resource, cancellationToken).ConfigureAwait(false);
+
+            if (exitCode != 0)
+            {
+                bool isLoginError = output.StartsWith("Please run 'az login'", StringComparison.CurrentCultureIgnoreCase);
+                bool isWinError = output.StartsWith(WinAzureCLIError, StringComparison.CurrentCultureIgnoreCase);
+                string pattter = "az:(.*)not found";
+                bool isOtherOsError = Regex.IsMatch(output, pattter);
+
+                if (isWinError || isOtherOsError)
+                {
+                    throw new CredentialUnavailableException(AzureCLINotInstalled);
+                }
+                else if (isLoginError)
+                {
+                    throw new CredentialUnavailableException(AzNotLogIn);
+                }
+
+                throw new AuthenticationFailedException(output);
+            }
+
+            byte[] byteArrary = Encoding.ASCII.GetBytes(output);
+
+            using MemoryStream stream = new MemoryStream(byteArrary);
+
+            return await AccessTokenUtilities.DeserializeAsync(async, stream, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected virtual async ValueTask<(string, int)> GetAzureCliAccesToken(bool async, string resource, CancellationToken cancellationToken)
+        {
+            string command = $"az account get-access-token --output json --resource {resource}";
```

Comment:
I wonder if we should build a test app that compiles into `az.exe` and use it from the test. This way we can write almost real integration test.

Reply:
We have a couple of features across this preview and the next which are executing processes. We're working on a more comprehensive mocking story for this.